source/served/extension.d

```diff
@@ -633,42 +656,41 @@ void delayedProjectActivation(WorkspaceD.Instance instance, string workspaceRoot
 		if (!loadedDub)
 			error("Exception starting dub: ", err);
 		else
-			trace("Started dub with root dependencies ", instance.get!DubComponent.rootDependencies);
+			trace("Started dub with root dependencies ", instance
+					.get!DubComponent.rootDependencies);
 	}
-	if (!loadedDub)
-	{
-		if (!disableDub)
-		{
-			error("Failed starting dub in ", root, " - falling back to fsworkspace");
-			proj.startupError(workspaceRoot, translate!"d.ext.dubFail"(instance.cwd, err ? err.msg : ""));
-		}
-		try
-		{
-			trace("Starting fsworkspace...");
 
-			instance.config.set("fsworkspace", "additionalPaths",
-					getPossibleSourceRoots(workspaceRoot));
-			if (!backend.attachEager(instance, "fsworkspace", err))
-				throw new Exception("Attach returned failure: " ~ err.msg);
-		}
-		catch (Exception e)
-		{
-			error(e);
-			proj.startupError(workspaceRoot, translate!"d.ext.fsworkspaceFail"(instance.cwd));
-		}
-	}
-	else
+	if (loadedDub)
 		didLoadDubProject();
 
-	trace("Started files provider for root ", root);
-
-	trace("Loaded Components for ", instance.cwd, ": ",
-			instance.instanceComponents.map!"a.info.name");
+	string ccdbPath = proj.config.d.ccdbPath;
+	if (!ccdbPath.length && !loadedDub)
+		ccdbPath = discoverCcdb(workspaceRoot);
+	if (ccdbPath.length)
+	{
+		trace("starting CCDB with ", ccdbPath);
+		instance.get!ClangCompilationDatabaseComponent.setDbPath(ccdbPath);
+	}
 
-	emitExtensionEvent!onAddedProject(instance, workspaceRoot, workspaceUri);
+	if (!loadedDub && !disableDub && !ccdbPath.length)
+	{
+		error("Failed starting dub or CCDB in ", root, " - falling back to fsworkspace");
+		proj.startupError(workspaceRoot, translate!"d.ext.dubFail"(instance.cwd, err ? err.msg : ""));
+	}
+	try
```

Comment:
you moved the `try` with fsworkspace out of the `if (!loadedDub)` check.

The error message was in a further `if (!disableDub)` to only show the error when DUB was enabled and actually attempted to be loaded. This should probably be changed to only load fsworkspace again if neither dub nor ccdb are loaded. (otherwise fsworkspace might override it)

This is probably the last thing I have to note for this PR, I still need to test out different projects of dub/fsworkspace/ccdb locally.

Reply:
Is it better with 43e2bbdb4833f68eda6ce7730d6d563df9784493?
I'm not too sure what the fallback logic should be. Do I need to add a `d.ext.ccdbFail` message? (and load the db in a try/catch)

Reply:
yeah that commit is better I think. A proper `ccdbFail` message would be nice for the user